File: backend/src/common/labor.ts
/**
 * Трудозатраты: штат и подряд на одном переделе (09_COSTING_AND_STAGES.md §5).
 *
 * Норма — свойство изделия, исполнитель — свойство заказа. Подряд это
 * ситуативное решение («не успеваем — обшивку отдали на сторону»), и влияет
 * оно на себестоимость конкретного заказа, а не на эталон в справочнике.
 */

export type LaborKindValue = 'STAFF' | 'CONTRACTOR';
export type WorkLocationValue = 'OUR_SHOP' | 'CONTRACTOR_SITE';
export type RateTypeValue = 'PER_HOUR' | 'PER_UNIT' | 'PER_KG' | 'PER_TON' | 'FIXED';
export type StageValue = 'CUTTING' | 'ASSEMBLY' | 'PAINTING';

/**
 * Системный дефолт — за штуку. Вес заполнен у 70 артикулов из 2 161 (3,2 %),
 * поэтому весовые ставки по умолчанию давали бы ноль на 97 % изделий,
 * а количество известно точно всегда.
 */
export const DEFAULT_RATE_TYPE: RateTypeValue = 'PER_UNIT';

export const RATE_TYPE_LABELS: Record<RateTypeValue, string> = {
  PER_HOUR: '₸ / час',
  PER_UNIT: '₸ / шт',
  PER_KG: '₸ / кг',
  PER_TON: '₸ / т',
  FIXED: '₸ за объём',
};

/** Ставки по весу невозможно посчитать без веса изделия */
export const WEIGHT_BASED_RATES: RateTypeValue[] = ['PER_KG', 'PER_TON'];

export class LaborConfigError extends Error {
  constructor(readonly code: string, message: string) {
    super(message);
    this.name = 'LaborConfigError';
  }
}

export interface LaborAssignment {
  id?: string;
  stage: StageValue;
  laborKind: LaborKindValue;
  /** Доля объёма передела: две строки по 0.5 — это «подряд вместе со штатом» */
  share: number;
  rateType: RateTypeValue;
  rate: number;
  /** Часы идут в загрузку цеха, только если работы выполняются у нас */
  countInShopHours: boolean;
  /** Норма изделия — нужна для PER_HOUR */
  workers?: number;
  hoursPerUnit?: number;
  /** Оценка часов для сдельного подряда, работающего в нашем цеху */
  plannedHours?: number | null;
  contractorId?: string | null;
  workCenterId?: string | null;
  /**
   * Измеренный факт подряда в единицах СВОЕЙ ставки (часы для PER_HOUR,
   * штуки для PER_UNIT, кг для PER_KG, тонны для PER_TON). Пока NULL —
   * объём считается плановым: qty × share.
   */
  actualQty?: number | null;
  /** Сумма, замороженная при приёмке работы: ставка потом может смениться */
  actualAmount?: number | null;
  /**
   * Доля работы, приходящаяся на ЭТУ позицию заказа, когда подряд заведён
   * на заказ целиком (0…1). Калькуляция считается по позициям, а «фикс
   * 500 000 ₸ за передел» и принятая сумма — величины абсолютные: без
   * разнесения они списались бы полностью в каждую позицию и заказ из двух
   * позиций стоил бы вдвое дороже.
   *
   * К сдельным плановым ставкам НЕ применяется: там объём и так берётся
   * из qty позиции, и сумма по позициям сходится сама.
   */
  allocationFactor?: number;
}

export interface LaborContext {
  qty: number;
  weightKg?: number | null;
}

export interface LaborLineCost {
  stage: StageValue;
  laborKind: LaborKindValue;
  share: number;
  rateType: RateTypeValue;
  rate: number;
  cost: number;
  manHours: number;
  countedInShopHours: boolean;
  contractorId?: string | null;
  workCenterId?: string | null;
}

const round2 = (n: number) => Math.round(n * 100) / 100;
const round3 = (n: number) => Math.round(n * 1000) / 1000;
const SHARE_TOLERANCE = 1e-4;

/** Доступен ли тип ставки для этого изделия (09 §5.2) */
export function rateTypeAvailable(rateType: RateTypeValue, weightKg?: number | null): boolean {
  return !WEIGHT_BASED_RATES.includes(rateType) || (weightKg ?? 0) > 0;
}

/**
 * Стоимость и трудоёмкость одной строки исполнения.
 *
 * Молчаливых нулей нет нигде: весовая ставка без веса и сдельный подряд
 * в нашем цеху без оценки часов — это ошибки, а не «ноль ₸» и «свободный цех».
 */
export function calcAssignmentCost(a: LaborAssignment, ctx: LaborContext): { cost: number; manHours: number } {
  const qty = Math.max(0, ctx.qty);
  const share = Math.max(0, a.share);
  const weightKg = ctx.weightKg ?? 0;

  // Доля заказ-уровневой работы, приходящаяся на эту позицию (см. поле)
  const alloc = a.allocationFactor != null && a.allocationFactor >= 0 ? a.allocationFactor : 1;

  // Измеренный объём уже выражен в единицах своей ставки — вес изделия
  // тогда не нужен: подрядчик сдал 3,2 тонны, а не «10 штук по 320 кг»
  const measuredRaw = a.actualQty != null && a.actualQty >= 0 ? a.actualQty : null;
  const measured = measuredRaw != null ? measuredRaw * alloc : null;

  if (WEIGHT_BASED_RATES.includes(a.rateType) && weightKg <= 0 && measuredRaw === null) {
    throw new LaborConfigError(
      'RATE_REQUIRES_WEIGHT',
      `Ставка ${RATE_TYPE_LABELS[a.rateType]} требует заполненного веса изделия`,
    );
  }

  const normManHours = round3((a.workers ?? 0) * (a.hoursPerUnit ?? 0) * qty * share);

  let cost: number;
  let manHours: number;

  switch (a.rateType) {
    case 'PER_HOUR':
      manHours = measured ?? normManHours;
      cost = round2(manHours * a.rate);
      break;
    case 'PER_UNIT':
      cost = round2((measured ?? qty * share) * a.rate);
      manHours = round3((a.plannedHours ?? 0) * alloc);
      break;
    case 'PER_KG':
      cost = round2((measured ?? qty * share * weightKg) * a.rate);
      manHours = round3((a.plannedHours ?? 0) * alloc);
      break;
    case 'PER_TON':
      cost = round2((measured ?? (qty * share * weightKg) / 1000) * a.rate);
      manHours = round3((a.plannedHours ?? 0) * alloc);
      break;
    case 'FIXED':
    default:
      // Фикс — сумма за весь передел: на позицию приходится своя доля
      cost = round2(a.rate * alloc);
      manHours = round3((a.plannedHours ?? 0) * alloc);
      break;
  }

  // Сумма, замороженная при приёмке, важнее пересчёта: платим то, что приняли,
  // даже если ставка в справочнике подрядчика с тех пор изменилась
  if (a.actualAmount != null && a.actualAmount >= 0) {
    cost = round2(a.actualAmount * alloc);
  }

  // Сдельная ставка не содержит часов. Если такие работы идут в нашем цеху,
  // без оценки планирование мощности покажет свободный участок, которого нет.
  if (a.countInShopHours && a.rateType !== 'PER_HOUR' && manHours <= 0) {
    throw new LaborConfigError(
      'SHOP_HOURS_ESTIMATE_REQUIRED',
      'Работы идут в нашем цеху по сдельной ставке — нужна оценка часов',
    );
  }

  return { cost, manHours };
}

/**
 * Проверять «сумма ровно 1» больше нельзя: штат не хранится строкой, он —
 * остаток от нормы (решение 23.08.2026). Осталась одна возможная ошибка:
 * подряд забрал больше целого передела, и тогда объём оплачен дважды.
 *
 * Недобор до единицы ошибкой НЕ является: это либо штатный остаток, либо
 * передел без нормы, где штату взяться неоткуда.
 */
export function validateShares(assignments: LaborAssignment[]): string | null {
  const byStage = new Map<StageValue, number>();
  for (const a of assignments) {
    if (a.share < 0) return `Доля не может быть отрицательной (${a.stage})`;
    byStage.set(a.stage, (byStage.get(a.stage) ?? 0) + a.share);
  }
  for (const [stage, sum] of byStage) {
    if (sum - 1 > SHARE_TOLERANCE) {
      return `Доли на переделе ${stage} дают ${round2(sum * 100)} % — объём посчитан дважды`;
    }
  }
  return null;
}

export interface LaborSummary {
  lines: LaborLineCost[];
  /** Трудозатраты цеха: штат плюс подряд, работающий у нас */
  shopManHours: number;
  staffCost: number;
  contractorCost: number;
  /** Из подряда — та часть, что выполняется на стороне */
  offsiteContractorCost: number;
  totalCost: number;
  byStage: Array<{ stage: StageValue; cost: number; manHours: number }>;
}

/** Сводка по всем строкам исполнения заказа */
export function summarizeLabor(assignments: LaborAssignment[], ctx: LaborContext): LaborSummary {
  const shareError = validateShares(assignments);
  if (shareError) throw new LaborConfigError('INVALID_SHARES', shareError);

  const lines: LaborLineCost[] = assignments.map((a) => {
    const { cost, manHours } = calcAssignmentCost(a, ctx);
    return {
      stage: a.stage,
      laborKind: a.laborKind,
      share: a.share,
      rateType: a.rateType,
      rate: a.rate,
      cost,
      manHours,
      countedInShopHours: a.countInShopHours,
      contractorId: a.contractorId ?? null,
      workCenterId: a.workCenterId ?? null,
    };
  });

  const sum = (pred: (l: LaborLineCost) => boolean, field: 'cost' | 'manHours') =>
    lines.filter(pred).reduce((s, l) => s + l[field], 0);

  const byStageMap = new Map<StageValue, { cost: number; manHours: number }>();
  for (const l of lines) {
    const cur = byStageMap.get(l.stage) ?? { cost: 0, manHours: 0 };
    byStageMap.set(l.stage, { cost: cur.cost + l.cost, manHours: cur.manHours + l.manHours });
  }

  return {
    lines,
    shopManHours: round3(sum((l) => l.countedInShopHours, 'manHours')),
    staffCost: round2(sum((l) => l.laborKind === 'STAFF', 'cost')),
    contractorCost: round2(sum((l) => l.laborKind === 'CONTRACTOR', 'cost')),
    offsiteContractorCost: round2(
      sum((l) => l.laborKind === 'CONTRACTOR' && !l.countedInShopHours, 'cost'),
    ),
    totalCost: round2(sum(() => true, 'cost')),
    byStage: [...byStageMap.entries()].map(([stage, v]) => ({
      stage,
      cost: round2(v.cost),
      manHours: round3(v.manHours),
    })),
  };
}

/**
 * Строки исполнения по умолчанию — целиком штат по нормам изделия.
 * Подряд появляется только явным решением человека.
 */
export function defaultStaffAssignments(
  norms: Array<{ stage: StageValue; workers: number; hoursPerUnit: number; hourlyRate: number; workCenterId?: string | null }>,
): LaborAssignment[] {
  return norms.map((n) => ({
    stage: n.stage,
    laborKind: 'STAFF',
    share: 1,
    rateType: 'PER_HOUR',
    rate: n.hourlyRate,
    countInShopHours: true,
    workers: n.workers,
    hoursPerUnit: n.hoursPerUnit,
    workCenterId: n.workCenterId ?? null,
  }));
}

File: backend-go/internal/warehouse/labor.go
// Перенос backend/src/common/labor.ts — трудозатраты: штат и подряд на
// одном переделе. Норма — свойство изделия, исполнитель — свойство заказа.
package warehouse

import (
	"math"
	"strconv"
)

func trimFloat(n float64) string { return strconv.FormatFloat(n, 'f', -1, 64) }

type LaborKind string

const (
	LaborKindStaff      LaborKind = "STAFF"
	LaborKindContractor LaborKind = "CONTRACTOR"
)

type RateType string

const (
	RateTypePerHour RateType = "PER_HOUR"
	RateTypePerUnit RateType = "PER_UNIT"
	RateTypePerKg   RateType = "PER_KG"
	RateTypePerTon  RateType = "PER_TON"
	RateTypeFixed   RateType = "FIXED"
)

var weightBasedRates = map[RateType]bool{RateTypePerKg: true, RateTypePerTon: true}

var rateTypeLabels = map[RateType]string{
	RateTypePerHour: "₸ / час", RateTypePerUnit: "₸ / шт", RateTypePerKg: "₸ / кг",
	RateTypePerTon: "₸ / т", RateTypeFixed: "₸ за объём",
}

// LaborConfigError — перенос LaborConfigError (плейн Error, не HttpException
// — тот же класс поведения, что CostingConfigError/BatchSelectionError).
type LaborConfigError struct {
	Code    string
	Message string
}

func (e *LaborConfigError) Error() string { return e.Message }

type Stage string

const (
	StageCutting  Stage = "CUTTING"
	StageAssembly Stage = "ASSEMBLY"
	StagePainting Stage = "PAINTING"
)

// LaborAssignment — перенос LaborAssignment.
type LaborAssignment struct {
	ID               string
	Stage            Stage
	LaborKind        LaborKind
	Share            float64
	RateType         RateType
	Rate             float64
	CountInShopHours bool
	Workers          float64
	HoursPerUnit     float64
	PlannedHours     *float64
	ContractorID     *string
	WorkCenterID     *string
	ActualQty        *float64
	ActualAmount     *float64
	AllocationFactor *float64
}

type LaborContext struct {
	Qty      float64
	WeightKg float64
}

type LaborLineCost struct {
	Stage              Stage     `json:"stage"`
	LaborKind          LaborKind `json:"laborKind"`
	Share              float64   `json:"share"`
	RateType           RateType  `json:"rateType"`
	Rate               float64   `json:"rate"`
	Cost               float64   `json:"cost"`
	ManHours           float64   `json:"manHours"`
	CountedInShopHours bool      `json:"countedInShopHours"`
	ContractorID       *string   `json:"contractorId"`
	WorkCenterID       *string   `json:"workCenterId"`
}

const shareTolerance = 1e-4

// RateTypeAvailable — доступен ли тип ставки для этого изделия (09 §5.2).
func RateTypeAvailable(rateType RateType, weightKg float64) bool {
	return !weightBasedRates[rateType] || weightKg > 0
}

// CalcAssignmentCost — стоимость и трудоёмкость одной строки исполнения.
// Молчаливых нулей нет нигде: весовая ставка без веса и сдельный подряд в
// нашем цеху без оценки часов — это ошибки, не «ноль» и «свободный цех».
func CalcAssignmentCost(a LaborAssignment, ctx LaborContext) (cost, manHours float64, err error) {
	qty := math.Max(0, ctx.Qty)
	share := math.Max(0, a.Share)
	weightKg := ctx.WeightKg

	alloc := 1.0
	if a.AllocationFactor != nil && *a.AllocationFactor >= 0 {
		alloc = *a.AllocationFactor
	}

	var measured *float64
	if a.ActualQty != nil && *a.ActualQty >= 0 {
		v := *a.ActualQty * alloc
		measured = &v
	}

	if weightBasedRates[a.RateType] && weightKg <= 0 && a.ActualQty == nil {
		return 0, 0, &LaborConfigError{
			Code:    "RATE_REQUIRES_WEIGHT",
			Message: "Ставка " + rateTypeLabels[a.RateType] + " требует заполненного веса изделия",
		}
	}

	normManHours := round3(a.Workers * a.HoursPerUnit * qty * share)
	plannedHours := 0.0
	if a.PlannedHours != nil {
		plannedHours = *a.PlannedHours
	}

	switch a.RateType {
	case RateTypePerHour:
		if measured != nil {
			manHours = *measured
		} else {
			manHours = normManHours
		}
		cost = round2(manHours * a.Rate)
	case RateTypePerUnit:
		volume := qty * share
		if measured != nil {
			volume = *measured
		}
		cost = round2(volume * a.Rate)
		manHours = round3(plannedHours * alloc)
	case RateTypePerKg:
		volume := qty * share * weightKg
		if measured != nil {
			volume = *measured
		}
		cost = round2(volume * a.Rate)
		manHours = round3(plannedHours * alloc)
	case RateTypePerTon:
		volume := (qty * share * weightKg) / 1000
		if measured != nil {
			volume = *measured
		}
		cost = round2(volume * a.Rate)
		manHours = round3(plannedHours * alloc)
	case RateTypeFixed:
		fallthrough
	default:
		cost = round2(a.Rate * alloc)
		manHours = round3(plannedHours * alloc)
	}

	// Сумма, замороженная при приёмке, важнее пересчёта
	if a.ActualAmount != nil && *a.ActualAmount >= 0 {
		cost = round2(*a.ActualAmount * alloc)
	}

	if a.CountInShopHours && a.RateType != RateTypePerHour && manHours <= 0 {
		return 0, 0, &LaborConfigError{
			Code:    "SHOP_HOURS_ESTIMATE_REQUIRED",
			Message: "Работы идут в нашем цеху по сдельной ставке — нужна оценка часов",
		}
	}

	return cost, manHours, nil
}

// ValidateShares — сумма долей на переделе не может превышать единицу
// (объём посчитан дважды). Недобор до единицы ошибкой не является.
func ValidateShares(assignments []LaborAssignment) string {
	byStage := map[Stage]float64{}
	for _, a := range assignments {
		if a.Share < 0 {
			return "Доля не может быть отрицательной (" + string(a.Stage) + ")"
		}
		byStage[a.Stage] += a.Share
	}
	for stage, sum := range byStage {
		if sum-1 > shareTolerance {
			return "Доли на переделе " + string(stage) + " дают " + trimFloat(round2(sum*100)) + " % — объём посчитан дважды"
		}
	}
	return ""
}

type LaborSummary struct {
	Lines                 []LaborLineCost    `json:"lines"`
	ShopManHours          float64            `json:"shopManHours"`
	StaffCost             float64            `json:"staffCost"`
	ContractorCost        float64            `json:"contractorCost"`
	OffsiteContractorCost float64            `json:"offsiteContractorCost"`
	TotalCost             float64            `json:"totalCost"`
	ByStage               []StageCostSummary `json:"byStage"`
}

type StageCostSummary struct {
	Stage    Stage   `json:"stage"`
	Cost     float64 `json:"cost"`
	ManHours float64 `json:"manHours"`
}

// SummarizeLabor — сводка по всем строкам исполнения заказа.
func SummarizeLabor(assignments []LaborAssignment, ctx LaborContext) (LaborSummary, error) {
	if errMsg := ValidateShares(assignments); errMsg != "" {
		return LaborSummary{}, &LaborConfigError{Code: "INVALID_SHARES", Message: errMsg}
	}

	lines := make([]LaborLineCost, len(assignments))
	for i, a := range assignments {
		cost, manHours, err := CalcAssignmentCost(a, ctx)
		if err != nil {
			return LaborSummary{}, err
		}
		lines[i] = LaborLineCost{
			Stage: a.Stage, LaborKind: a.LaborKind, Share: a.Share, RateType: a.RateType, Rate: a.Rate,
			Cost: cost, ManHours: manHours, CountedInShopHours: a.CountInShopHours,
			ContractorID: a.ContractorID, WorkCenterID: a.WorkCenterID,
		}
	}

	shopManHours, staffCost, contractorCost, offsiteContractorCost, totalCost := 0.0, 0.0, 0.0, 0.0, 0.0
	byStageMap := map[Stage]*StageCostSummary{}
	var stageOrder []Stage
	for _, l := range lines {
		if l.CountedInShopHours {
			shopManHours += l.ManHours
		}
		if l.LaborKind == LaborKindStaff {
			staffCost += l.Cost
		}
		if l.LaborKind == LaborKindContractor {
			contractorCost += l.Cost
			if !l.CountedInShopHours {
				offsiteContractorCost += l.Cost
			}
		}
		totalCost += l.Cost
		s, ok := byStageMap[l.Stage]
		if !ok {
			s = &StageCostSummary{Stage: l.Stage}
			byStageMap[l.Stage] = s
			stageOrder = append(stageOrder, l.Stage)
		}
		s.Cost += l.Cost
		s.ManHours += l.ManHours
	}
	byStage := make([]StageCostSummary, 0, len(stageOrder))
	for _, s := range stageOrder {
		v := byStageMap[s]
		byStage = append(byStage, StageCostSummary{Stage: v.Stage, Cost: round2(v.Cost), ManHours: round3(v.ManHours)})
	}

	return LaborSummary{
		Lines: lines, ShopManHours: round3(shopManHours), StaffCost: round2(staffCost),
		ContractorCost: round2(contractorCost), OffsiteContractorCost: round2(offsiteContractorCost),
		TotalCost: round2(totalCost), ByStage: byStage,
	}, nil
}

// DefaultStaffAssignments — строки исполнения по умолчанию: целиком штат по нормам изделия.
type NormLike struct {
	Stage        Stage
	Workers      float64
	HoursPerUnit float64
	HourlyRate   float64
	WorkCenterID *string
}

func DefaultStaffAssignments(norms []NormLike) []LaborAssignment {
	out := make([]LaborAssignment, len(norms))
	for i, n := range norms {
		out[i] = LaborAssignment{
			Stage: n.Stage, LaborKind: LaborKindStaff, Share: 1, RateType: RateTypePerHour, Rate: n.HourlyRate,
			CountInShopHours: true, Workers: n.Workers, HoursPerUnit: n.HoursPerUnit, WorkCenterID: n.WorkCenterID,
		}
	}
	return out
}
